### DISCUSSION TITLE
The choice loss function in the spike rate coding

### DISCUSSION THREAD

**Discussion by HuisheepLuo** [jeshraghian/snntorch#81] [Q&A]
Hello,

I try to build a spike rate code based network following the tutorials, and I have some problems when I call the loss function.
It doesn't seem to converge when I call loss function.

`loss_fn = SF.ce_count_loss()` 

then,
``` 
for i, (data, targets) in enumerate(iter(train_loader)):
        data = data.to(device)
        targets = targets.to(device)
        spike_data = spikegen.rate(data, num_steps=num_steps)
        net.train()
        spk_rec, mem_rec = net(spike_data)
        loss_val = loss_fn(spk_rec, targets)
```

The loss would not change during iteration.

The situation get better when I pass the parameter `mem_rec` instead of `spk_rec` as the input of `loss_fn`. However, I don't think it's a good idea to use membrane potentials as loss count, since I try to build the SNN with spikes input and output.

The same problem appeared when I change the loss function to MSE.

Thanks and looking for your response.

**Comment by jeshraghian** [comment]
The most likely problem is that the spike patterns in `spk_rec` do not change after each update. 

Try to plot a raster of the output spikes. If there is little/no spiking, you may need to decrease `threshold` when defining the neuron.
Alternatively, you may need to simulate for more time steps to give the network a longer window for spiking behavior to vary. A larger learning rate may also help.

**Reply by HuisheepLuo** [reply]
Thanks for your answer.
It works well when I add the `spike_grad` into the lif layer.The non-convergent situation is only when I don't add `spike_grad` and use `ce_count_loss()`. Maybe it's kind of vanishing gradient.

**Reply by jeshraghian** [reply]
Nice! Yeah, without `spike_grad`, the gradient for U>threshold is dS/dU=1 if , and if U < threshold, then dS/dU=0.

This means there are a lot of 0's, and the vanishing gradient problem can easily arise.
